conda/gateways/disk/lock.py

```diff
@@ -30,13 +31,16 @@ def _lock_noop(fd):
     def _lock_impl(fd):  # type: ignore
         tell = fd.tell()
         fd.seek(LOCK_BYTE)
-        msvcrt.locking(fd.fileno(), msvcrt.LK_LOCK, 1)  # type: ignore
         try:
-            fd.seek(tell)
-            yield
-        finally:
-            fd.seek(LOCK_BYTE)
-            msvcrt.locking(fd.fileno(), msvcrt.LK_UNLCK, 1)  # type: ignore
+            msvcrt.locking(fd.fileno(), msvcrt.LK_LOCK, 1)  # type: ignore
+            try:
+                fd.seek(tell)
+                yield
+            finally:
+                fd.seek(LOCK_BYTE)
+                msvcrt.locking(fd.fileno(), msvcrt.LK_UNLCK, 1)  # type: ignore
+        except OSError as e:
+            raise LockError("Failed to acquire lock.") from e
```

Comment:
The reason the test is failing is because we are catching ALL `OSErrors` that occur, both those raised during locking but also those raised within the context.

We need the following to raise an `OSError` but currently this will raise a `LockError` even if the locking was successful:
```python
with lock(...):
    raise OSError
```